### ISSUE TITLE
Support for String Concatenation after Java 9 is missing.

### ISSUE THREAD

**Issue by michaelphop** [pascal-lab/Tai-e#50]
public class Foo {

    @Override
    public String toString() {
        return "str";
    }
}

public class Main {
 
    public static void main(String[] args) {
       
        System.out.println("some print" + new Foo());
    }
}

In the above example, when string concat is being performed, the call to ```toString``` of ```Foo``` is not in "call-edges.txt" (running with ```only-app:false```)

**Comment by zhangt2333** [comment]
The call edge `<java.lang.String: java.lang.String valueOf(java.lang.Object)>/java.lang.Object.toString/0	<Test$Foo: java.lang.String toString()>` exists when Tai-e options are: 

```
"-pp"
"-cp" "path/to/Main"
"-m" "Main"
"-a" "pta=implicit-entries:false;only-app:false"
"-a" "cg=dump-call-edges:true;"
```

**Comment by michaelphop** [comment]
I put these options and still this edge doesn't exist

**Comment by zhangt2333** [comment]
I offer a reproduction in https://github.com/pascal-lab/Tai-e/tree/issues/50.

See https://github.com/pascal-lab/Tai-e/commit/5f39e00e4e1f727115858c21baf6aad8ea00322f.

**Comment by michaelphop** [comment]
I run with these configurations:
```
-cp
Test.jar
-m
org.example77.Main
-pp
-scope
ALL
-a
pta=handle-invokedynamic:true;cs:2-type;dump:false;reflection-inference:solar;only-app:false;handle-invokedynamic:true;merge-string-objects:true;distinguish-string-constants:all
-a
cg=dump-call-edges:true
```

**Comment by michaelphop** [comment]
attached the call-edges.txt
[call-edges.txt](https://github.com/pascal-lab/Tai-e/files/12215398/call-edges.txt)

**Comment by michaelphop** [comment]
package org.example77;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;



public class Main {

    public static void main(String[] args) throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, IllegalAccessException, InstantiationException, IOException {
        System.out.println("str" + new Foo());
    }
}



package org.example77;

public class Foo {

    @Override
    public String toString() {
        return "str";
    }
}

**Comment by michaelphop** [comment]
the run arguments provided above

**Comment by michaelphop** [comment]
also the call-edges.txt

**Comment by zhangt2333** [comment]
See https://github.com/pascal-lab/Tai-e/tree/issues/50.

Anything goes well.

**Comment by michaelphop** [comment]
Yes I saw this, its not in call-edges.txt (attached above)

**Comment by zhangt2333** [comment]
How did you get your "call-edges.txt" and can you provide me with a reproducible code **like I did** (an archive file or a GitHub repository)?

**Comment by michaelphop** [comment]
attached my jar file:
[Test.zip](https://github.com/pascal-lab/Tai-e/files/12216234/Test.zip)

The call-edges.txt I get after analysis finished from: "output/call-edges.txt"

**Comment by zhangt2333** [comment]
All right, It's reproduced.

I will delve deeper into it tomorrow.

**Comment by michaelphop** [comment]
Thank you

**Comment by michaelphop** [comment]
Is there a requirement for my java code to be compiled with specific jdk?
I am asking because when I compiled my ```Test.jar``` with java8 the edge was seen, but when compiled with java17, the problem occurred (missing edge of ```toString```)

**Comment by zhangt2333** [comment]
In Java9, String Concatenation is replaced by calling `StringConcatFactory.makeConcatWithConstants` with `InvokeDynamic` instruction, instead of `StringBuilder` (see https://openjdk.org/jeps/280).

JDK8 IR:
```java
public class Main extends java.lang.Object {
...
    public static void main(java.lang.String[] r6) {
        java.io.PrintStream $r1;
        java.lang.StringBuilder $r0, $r2, $r4;
        java.lang.String %stringconst0, $r5;
        Foo $r3;
        [0@L3] $r1 = <java.lang.System: java.io.PrintStream out>;
        [1@L3] $r0 = new java.lang.StringBuilder;
        [2@L3] invokespecial $r0.<java.lang.StringBuilder: void <init>()>();
        [3@L3] %stringconst0 = "some print";
        [4@L3] $r2 = invokevirtual $r0.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>(%stringconst0);
        [5@L3] $r3 = new Foo;
        [6@L3] invokespecial $r3.<Foo: void <init>()>();
        [7@L3] $r4 = invokevirtual $r2.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.Object)>($r3);
        [8@L3] $r5 = invokevirtual $r4.<java.lang.StringBuilder: java.lang.String toString()>();
        [9@L3] invokevirtual $r1.<java.io.PrintStream: void println(java.lang.String)>($r5);
        [10@L4] return;
    }
}
```

JDK17 IR:
```java
public class Main extends java.lang.Object {
...
    public static void main(java.lang.String[] r3) {
        java.io.PrintStream $r1;
        Foo $r0;
        java.lang.String $r2;
        [0@L3] $r1 = <java.lang.System: java.io.PrintStream out>;
        [1@L3] $r0 = new Foo;
        [2@L3] invokespecial $r0.<Foo: void <init>()>();
        [3@L3] $r2 = invokedynamic <java.lang.invoke.StringConcatFactory: java.lang.invoke.CallSite makeConcatWithConstants(java.lang.invoke.MethodHandles$Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.String,java.lang.Object[])> "makeConcatWithConstants" <MethodType: java.lang.String (Foo)>["some print"]($r0);
        [4@L3] invokevirtual $r1.<java.io.PrintStream: void println(java.lang.String)>($r2);
        [5@L4] return;
    }

}
```

We are working on getting Tai-e support to handle this feature. As a temporary alternative, you might consider analysing the JDK8 program first

**Comment by zhangt2333** [comment]
Fixed in latest commits ba400b4c1fa8bb6d20e82f113690d6cd0e7a2b43...d52619a62099cf8f9c8732e2c19c7c45970f2e9d. 

Issue-related tests: https://github.com/pascal-lab/Tai-e/commit/e03a392ad660541ce3ca8bb6050346d1053adfb7.
